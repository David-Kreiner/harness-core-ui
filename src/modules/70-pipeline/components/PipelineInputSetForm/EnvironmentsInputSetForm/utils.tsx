/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { get, isEmpty } from 'lodash-es'

import { RUNTIME_INPUT_VALUE, SelectOption } from '@harness/uicore'

import { DeploymentStageConfig } from 'services/cd-ng'
import { StageElementWrapperConfig } from 'services/pipeline-ng'

import { isValueRuntimeInput } from '@common/utils/utils'

import { StageType } from '@pipeline/utils/stageHelpers'
import { DeploymentStageElementConfig } from '@pipeline/utils/pipelineTypes'

export interface PropagateSelectOption extends SelectOption {
  infraLabel?: string
}

export function getPropagateStageOptions(
  stages: StageElementWrapperConfig[],
  stageIndex: number,
  templateTypes: {
    [key: string]: string
  }
): PropagateSelectOption[] {
  if (stages.length && stageIndex > 0) {
    const getStagesAllowedforPropagate = (stageItem: StageElementWrapperConfig): boolean => {
      if (stageItem.stage?.template) {
        const isAlreadyPropagatingValue = !!(stageItem?.stage?.template?.templateInputs?.spec as DeploymentStageConfig)
          ?.environment?.useFromStage?.stage
        const stageType = get(templateTypes, stageItem.stage.template.templateRef)
        return stageType === StageType.DEPLOY && !isAlreadyPropagatingValue
      } else {
        const isAlreadyPropagatingValue = !!(stageItem?.stage?.spec as DeploymentStageConfig)?.environment?.useFromStage
          ?.stage
        const stageType = stageItem?.stage?.type
        const isSingleEnvEmpty = isEmpty(
          (stageItem.stage as DeploymentStageElementConfig)?.spec?.environment?.environmentRef
        )
        return stageType === StageType.DEPLOY && !isSingleEnvEmpty && !isAlreadyPropagatingValue
      }
    }

    const stageWithEnvV2 = stages.slice(0, stageIndex).filter(getStagesAllowedforPropagate)
    return stageWithEnvV2.map(stageItem => {
      if (stageItem.stage?.template) {
        return {
          label: `Stage [${stageItem.stage?.name}] - Template [${stageItem.stage.template.templateRef}]`,
          value: stageItem.stage?.identifier as string,
          infraLabel: `Stage [${stageItem.stage?.name}] - Template [${stageItem.stage.template.templateRef}]`
        }
      } else {
        const singleEnvironmentRef = (stageItem.stage as DeploymentStageElementConfig)?.spec?.environment
          ?.environmentRef
        const environmentLabel = `Environment [${singleEnvironmentRef}]`

        const singleInfrastructureRef = isValueRuntimeInput(
          // eslint-disable-next-line @typescript-eslint/no-explicit-any
          (stageItem.stage as DeploymentStageElementConfig)?.spec?.environment?.infrastructureDefinitions as any
        )
          ? RUNTIME_INPUT_VALUE
          : (stageItem.stage as DeploymentStageElementConfig)?.spec?.environment?.infrastructureDefinitions?.[0]
              ?.identifier
        const infrastructureLabel = `Infrastructure [${singleInfrastructureRef}]`

        return {
          label: `Stage [${stageItem.stage?.name}] - ${environmentLabel}`,
          value: stageItem.stage?.identifier as string,
          infraLabel: `Stage [${stageItem.stage?.name}] - ${infrastructureLabel}`
        }
      }
    })
  }
  return []
}
