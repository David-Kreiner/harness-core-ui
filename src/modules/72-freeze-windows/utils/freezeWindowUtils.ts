/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import type { FreezeDetailedResponse } from 'services/cd-ng'

export const RECURRENCE = ['Daily', 'Weekly', 'Monthly', 'Annually'] as const
export const DOES_NOT_REPEAT = 'Does not repeat'

type Scope = Exclude<FreezeDetailedResponse['freezeScope'], 'unknown' | undefined>

export const scopeText: Record<Scope, string> = {
  account: 'Account',
  org: 'Organization',
  project: 'Project'
}
