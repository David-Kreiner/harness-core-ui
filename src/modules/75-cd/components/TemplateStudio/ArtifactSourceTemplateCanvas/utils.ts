import type { ArtifactType } from '@pipeline/components/ArtifactsSelection/ArtifactInterface'
import { ArtifactToConnectorMap } from '@pipeline/components/ArtifactsSelection/ArtifactHelper'

export interface Params {
  oldConnectorRef?: string
  newConnectorRef?: string
  artifactType: ArtifactType
}

export const isConnectorRefUpdated = (params: Params) => {
  const { oldConnectorRef, newConnectorRef, artifactType } = params
  const isConnectorRefApplicable = Boolean(ArtifactToConnectorMap[artifactType])

  return Boolean(isConnectorRefApplicable && oldConnectorRef && newConnectorRef && oldConnectorRef !== newConnectorRef)
}
