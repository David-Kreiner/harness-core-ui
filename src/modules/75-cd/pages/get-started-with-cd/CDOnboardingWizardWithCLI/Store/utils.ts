export const parseJSON = (str: string): object => {
  try {
    return JSON.parse(str)
  } catch (e) {
    return {}
  }
}
