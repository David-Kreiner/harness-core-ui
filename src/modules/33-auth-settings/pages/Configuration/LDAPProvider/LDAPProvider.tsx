/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Free Trial 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/05/PolyForm-Free-Trial-1.0.0.txt.
 */

import React, { Dispatch, SetStateAction, useState } from 'react'
import cx from 'classnames'
import { useParams } from 'react-router-dom'
import {
  Radio,
  Container,
  Collapse,
  Card,
  Text,
  Button,
  Popover,
  ButtonVariation,
  Utils,
  useConfirmationDialog
} from '@wings-software/uicore'
import { Color, FontVariation } from '@harness/design-system'
import { Menu, MenuItem } from '@blueprintjs/core'
import { useToaster } from '@common/components'
import { String, useStrings } from 'framework/strings'
import type { AccountPathProps } from '@common/interfaces/RouteInterfaces'
import {
  AuthenticationSettingsResponse,
  LDAPSettings,
  syncLdapGroupsPromise,
  useDeleteLdapSettings
} from 'services/cd-ng'
import { AuthenticationMechanisms } from '@auth-settings/constants/utils'
import RbacMenuItem from '@rbac/components/MenuItem/MenuItem'
import { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import type { PermissionRequest } from '@auth-settings/pages/Configuration/Configuration'
import { useFeature } from '@common/hooks/useFeatures'
import { FeatureIdentifier } from 'framework/featureStore/FeatureIdentifier'
import useRBACError from '@rbac/utils/useRBACError/useRBACError'
import { useFeatureFlag } from '@common/hooks/useFeatureFlag'
import { FeatureFlag } from '@common/featureFlags'
import { FeatureWarningTooltip } from '@common/components/FeatureWarning/FeatureWarningWithTooltip'
import useLdapTestConfigurationProvider from '@auth-settings/modals/LdapTestConfiguration/useLdapTestConfigurationProvider'
import css from './LDAPProvider.module.scss'
import cssConfiguration from '@auth-settings/pages/Configuration/Configuration.module.scss'

interface Props {
  authSettings: AuthenticationSettingsResponse
  refetchAuthSettings: () => void
  permissionRequest: PermissionRequest
  canEdit: boolean
  setUpdating: Dispatch<SetStateAction<boolean>>
}

const LDAPProvider: React.FC<Props> = ({ authSettings, refetchAuthSettings, permissionRequest, canEdit }) => {
  const { getRBACErrorMessage } = useRBACError()
  const { getString } = useStrings()
  const { showSuccess, showError } = useToaster()
  const { accountId } = useParams<AccountPathProps>()
  const ldapEnabledForAuth = authSettings.authenticationMechanism === AuthenticationMechanisms.LDAP
  const ldapSettings = authSettings.ngAuthSettings?.find(
    settings => settings.settingsType === AuthenticationMechanisms.LDAP
  ) as LDAPSettings | undefined

  const { enabled: featureEnabled } = useFeature({
    featureRequest: {
      featureName: FeatureIdentifier.LDAP_SUPPORT
    }
  })

  const [isLdapSyncInProgress, setIsLdapSyncInProgress] = useState<boolean>(false)

  const isNgLDAPFFEnabled = useFeatureFlag(FeatureFlag.NG_ENABLE_LDAP_CHECK)
  const onSuccess = (): void => {
    refetchAuthSettings()
  }

  const { openLdapTestModal } = useLdapTestConfigurationProvider({ onSuccess })

  const { mutate: deleteLdapSettings, loading: deletingLdapSettings } = useDeleteLdapSettings({
    queryParams: {
      accountIdentifier: accountId
    }
  })

  const { openDialog: confirmLdapSettingsDelete } = useConfirmationDialog({
    titleText: getString('authSettings.deleteLdapProvider'),
    contentText: (
      <String
        useRichText
        stringID="authSettings.deleteLdapProviderDescription"
        vars={{ displayName: ldapSettings?.displayName }}
      />
    ),
    confirmButtonText: getString('confirm'),
    cancelButtonText: getString('cancel'),
    onCloseDialog: async isConfirmed => {
      /* istanbul ignore else */ if (isConfirmed) {
        try {
          const deleted = await deleteLdapSettings()
          /* istanbul ignore else */ if (deleted) {
            refetchAuthSettings()
            showSuccess(getString('authSettings.ldapProviderDeleted'), 5000)
          }
        } catch (e) {
          /* istanbul ignore next */ showError(getRBACErrorMessage(e), 5000)
        }
      }
    }
  })

  if (!isNgLDAPFFEnabled) {
    return null
  }

  return (
    <Container margin="xlarge" background={Color.WHITE}>
      {ldapSettings ? (
        <Collapse
          isOpen={featureEnabled}
          collapseHeaderClassName={cx(cssConfiguration.collapseHeaderClassName, cssConfiguration.height60)}
          collapseClassName={cssConfiguration.collapseClassName}
          collapsedIcon="main-chevron-down"
          expandedIcon="main-chevron-up"
          heading={
            <Utils.WrapOptionalTooltip
              tooltip={
                !featureEnabled ? <FeatureWarningTooltip featureName={FeatureIdentifier.LDAP_SUPPORT} /> : undefined
              }
            >
              <Container margin={{ left: 'xlarge' }}>
                <Radio
                  checked={ldapEnabledForAuth}
                  font={{ weight: 'bold', size: 'normal' }}
                  color={Color.GREY_900}
                  label={getString('authSettings.loginViaLDAP')}
                  onChange={() => {
                    openLdapTestModal()
                  }}
                  disabled={!featureEnabled || !canEdit}
                />
              </Container>
            </Utils.WrapOptionalTooltip>
          }
        >
          <Container padding={{ bottom: 'large' }}>
            <Card className={css.card}>
              <Container margin={{ left: 'xlarge' }}>
                <Radio font={{ weight: 'bold', size: 'normal' }} checked={ldapEnabledForAuth} readOnly />
              </Container>
              <Text font={{ variation: FontVariation.CARD_TITLE }} className={css.flexGrowFull}>
                {ldapSettings.displayName}
              </Text>
              <Button
                text={getString('test')}
                variation={ButtonVariation.SECONDARY}
                onClick={() => {
                  openLdapTestModal()
                }}
                data-testid="ldap-config-test"
              />
              <Popover
                interactionKind="click"
                position="left-top"
                content={
                  <Menu data-testid="ldap-popover-menu">
                    <MenuItem text={getString('edit')} disabled={!canEdit} />
                    <RbacMenuItem
                      text={getString('delete')}
                      onClick={confirmLdapSettingsDelete}
                      permission={{
                        ...permissionRequest,
                        permission: PermissionIdentifier.DELETE_AUTHSETTING
                      }}
                      disabled={deletingLdapSettings}
                      data-testid="ldap-delete-config"
                    />
                    <MenuItem
                      text={getString('authSettings.ldap.syncUserGroups')}
                      data-testid="ldap-group-sync"
                      onClick={async () => {
                        try {
                          setIsLdapSyncInProgress(true)
                          const { resource } = await syncLdapGroupsPromise({
                            queryParams: {
                              accountIdentifier: accountId
                            }
                          })
                          if (resource) {
                            showSuccess(getString('authSettings.ldap.syncUserGroupsResult.success'), 5000)
                          } else {
                            showError(getString('authSettings.ldap.syncUserGroupsResult.fail'), 5000)
                          }
                        } catch (e) {
                          showError(getString('authSettings.ldap.syncUserGroupsResult.fail'), 5000)
                        } finally {
                          setIsLdapSyncInProgress(false)
                        }
                      }}
                      disabled={!canEdit || isLdapSyncInProgress}
                    />
                  </Menu>
                }
              >
                <Button minimal icon="Options" data-testid="provider-button" variation={ButtonVariation.ICON} />
              </Popover>
            </Card>
          </Container>
        </Collapse>
      ) : (
        <Utils.WrapOptionalTooltip
          tooltip={!featureEnabled ? <FeatureWarningTooltip featureName={FeatureIdentifier.LDAP_SUPPORT} /> : undefined}
        >
          <Card className={css.cardWithRadioBtn}>
            <Container margin={{ left: 'xlarge', top: 'xsmall' }}>
              <Radio
                checked={false}
                font={{ weight: 'semi-bold', size: 'normal' }}
                color={Color.PRIMARY_7}
                label={getString('authSettings.ldap.addLdap')}
                disabled={!featureEnabled || !canEdit}
              />
            </Container>
          </Card>
        </Utils.WrapOptionalTooltip>
      )}
    </Container>
  )
}

export default LDAPProvider
