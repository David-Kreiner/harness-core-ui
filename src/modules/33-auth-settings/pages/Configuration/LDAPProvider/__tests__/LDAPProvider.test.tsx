/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { act, fireEvent, render, waitFor } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import type { AuthenticationSettingsResponse } from 'services/cd-ng'
import * as cdngServices from 'services/cd-ng'
import LDAPProvider from '../LDAPProvider'
import { mockAuthSettingsResponse, permissionRequest } from './mock'

const refetchAuthSettings = jest.fn()
const mockDispatch = jest.fn()
const mockGroupSync = jest.fn().mockReturnValue(Promise.resolve(true))
const mockLdapLoginTest = jest.fn().mockReturnValue(
  Promise.resolve({
    resource: {
      status: 'SUCCESS'
    }
  })
)
const mockLdapLoginTestFail = jest.fn().mockReturnValue(
  Promise.resolve({
    resource: {
      status: 'FAILURE',
      message: 'Invalid Credentials'
    }
  })
)

const mockLdapLoginDelegateFailure = jest.fn().mockReturnValue(
  Promise.resolve({
    resource: {
      status: 'FAILURE',
      message: 'Delegate not found'
    }
  })
)
const mockUpdateAuthMechanism = jest.fn().mockReturnValue(
  Promise.resolve({
    resource: {
      status: 'SUCCESS'
    }
  })
)
const mockUpdateAuthMechanismFailure = jest.fn().mockReturnValue(
  Promise.resolve({
    responseMessages: [
      {
        message: 'Not Allowed'
      }
    ]
  })
)

describe('LDAP Provider', () => {
  beforeEach(() => {
    jest.clearAllMocks()
  })
  test('LDAP Provider Snapshot', async () => {
    const { container } = render(
      <TestWrapper pathParams={{ accountId: 'testAcc' }} defaultFeatureFlagValues={{ NG_ENABLE_LDAP_CHECK: true }}>
        <LDAPProvider
          authSettings={mockAuthSettingsResponse as AuthenticationSettingsResponse}
          canEdit={true}
          refetchAuthSettings={refetchAuthSettings}
          permissionRequest={permissionRequest}
          setUpdating={mockDispatch}
        />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })
  test('LDAP Provider Snapshot with no data', async () => {
    const { container } = render(
      <TestWrapper pathParams={{ accountId: 'testAcc' }} defaultFeatureFlagValues={{ NG_ENABLE_LDAP_CHECK: true }}>
        <LDAPProvider
          authSettings={[] as AuthenticationSettingsResponse}
          canEdit={true}
          refetchAuthSettings={refetchAuthSettings}
          permissionRequest={permissionRequest}
          setUpdating={mockDispatch}
        />
      </TestWrapper>
    )
    expect(container).toMatchSnapshot()
  })
  test('LDAP Provider delete', async () => {
    jest
      .spyOn(cdngServices, 'useDeleteLdapSettings')
      .mockReturnValue({ mutate: jest.fn(() => Promise.resolve(true)) } as any)
    const { getByTestId, getByText } = render(
      <TestWrapper pathParams={{ accountId: 'testAcc' }} defaultFeatureFlagValues={{ NG_ENABLE_LDAP_CHECK: true }}>
        <LDAPProvider
          authSettings={mockAuthSettingsResponse as AuthenticationSettingsResponse}
          canEdit={true}
          refetchAuthSettings={refetchAuthSettings}
          permissionRequest={permissionRequest}
          setUpdating={mockDispatch}
        />
      </TestWrapper>
    )
    const ldapPopoverEl = getByTestId('provider-button')
    expect(ldapPopoverEl).not.toBeNull()
    act(() => {
      ldapPopoverEl && fireEvent.click(ldapPopoverEl)
    })
    await waitFor(() => getByTestId('ldap-popover-menu'))
    const deleteConfigBtn = getByTestId('ldap-delete-config')
    expect(deleteConfigBtn).not.toBeNull()

    await act(async () => {
      deleteConfigBtn && fireEvent.click(deleteConfigBtn)
    })
    await waitFor(() => getByText('confirm'))
    const confirmDeleteBtn = getByText('confirm')
    await act(async () => {
      confirmDeleteBtn && fireEvent.click(confirmDeleteBtn)
    })
    await waitFor(() => getByText('authSettings.ldapProviderDeleted'))
  })
  test('LDAP Provider enable config', async () => {
    jest.spyOn(cdngServices, 'usePostLdapLoginTest').mockImplementation(
      () =>
        ({
          mutate: mockLdapLoginTest
        } as any)
    )
    jest.spyOn(cdngServices, 'useUpdateAuthMechanism').mockImplementation(
      () =>
        ({
          mutate: mockUpdateAuthMechanism
        } as any)
    )
    const { getByTestId, getByText } = render(
      <TestWrapper pathParams={{ accountId: 'testAcc' }} defaultFeatureFlagValues={{ NG_ENABLE_LDAP_CHECK: true }}>
        <LDAPProvider
          authSettings={mockAuthSettingsResponse as AuthenticationSettingsResponse}
          canEdit={true}
          refetchAuthSettings={refetchAuthSettings}
          permissionRequest={permissionRequest}
          setUpdating={mockDispatch}
        />
      </TestWrapper>
    )

    const testConfigBtn = getByTestId('ldap-config-test')
    act(() => {
      testConfigBtn && fireEvent.click(testConfigBtn)
    })
    await waitFor(() => expect(getByText('authSettings.ldap.verifyAndEnableConfig')).not.toBeNull())
    const enableLdapConfigBtn = getByTestId('enable-ldap-config')
    await act(async () => {
      enableLdapConfigBtn && fireEvent.click(enableLdapConfigBtn)
    })
    const emailField = document.querySelector('[name="email"]')
    const passwordField = document.querySelector('[name="password"]')
    await act(async () => {
      emailField && fireEvent.change(emailField, { target: { value: 'test@harness.io' } })
      passwordField && fireEvent.change(passwordField, { target: { value: 'test@1234' } })
    })
    await act(async () => {
      enableLdapConfigBtn && fireEvent.click(enableLdapConfigBtn)
    })
    await waitFor(() => expect(getByText('authSettings.ldap.authChangeSuccessful')).not.toBeNull())
  })
  test('LDAP Provider enable config test failure', async () => {
    jest.spyOn(cdngServices, 'usePostLdapLoginTest').mockImplementation(
      () =>
        ({
          mutate: mockLdapLoginDelegateFailure
        } as any)
    )
    jest.spyOn(cdngServices, 'useUpdateAuthMechanism').mockImplementation(
      () =>
        ({
          mutate: mockUpdateAuthMechanism
        } as any)
    )
    const { getByTestId, getByText } = render(
      <TestWrapper pathParams={{ accountId: 'testAcc' }} defaultFeatureFlagValues={{ NG_ENABLE_LDAP_CHECK: true }}>
        <LDAPProvider
          authSettings={mockAuthSettingsResponse as AuthenticationSettingsResponse}
          canEdit={true}
          refetchAuthSettings={refetchAuthSettings}
          permissionRequest={permissionRequest}
          setUpdating={mockDispatch}
        />
      </TestWrapper>
    )

    const testConfigBtn = getByTestId('ldap-config-test')
    act(() => {
      testConfigBtn && fireEvent.click(testConfigBtn)
    })
    await waitFor(() => expect(getByText('authSettings.ldap.verifyAndEnableConfig')).not.toBeNull())
    const enableLdapConfigBtn = getByTestId('enable-ldap-config')
    await act(async () => {
      enableLdapConfigBtn && fireEvent.click(enableLdapConfigBtn)
    })
    const emailField = document.querySelector('[name="email"]')
    const passwordField = document.querySelector('[name="password"]')
    await act(async () => {
      emailField && fireEvent.change(emailField, { target: { value: 'test@harness.io' } })
      passwordField && fireEvent.change(passwordField, { target: { value: 'test@1234' } })
    })
    await act(async () => {
      enableLdapConfigBtn && fireEvent.click(enableLdapConfigBtn)
    })
    await waitFor(() => expect(getByText('Delegate not found')).not.toBeNull())
  })
  test('LDAP Provider enable config auth failure', async () => {
    jest.spyOn(cdngServices, 'usePostLdapLoginTest').mockImplementation(
      () =>
        ({
          mutate: mockLdapLoginTest
        } as any)
    )
    jest.spyOn(cdngServices, 'useUpdateAuthMechanism').mockImplementation(
      () =>
        ({
          mutate: mockUpdateAuthMechanismFailure
        } as any)
    )
    const { getByTestId, getByText } = render(
      <TestWrapper pathParams={{ accountId: 'testAcc' }} defaultFeatureFlagValues={{ NG_ENABLE_LDAP_CHECK: true }}>
        <LDAPProvider
          authSettings={mockAuthSettingsResponse as AuthenticationSettingsResponse}
          canEdit={true}
          refetchAuthSettings={refetchAuthSettings}
          permissionRequest={permissionRequest}
          setUpdating={mockDispatch}
        />
      </TestWrapper>
    )

    const testConfigBtn = getByTestId('ldap-config-test')
    act(() => {
      testConfigBtn && fireEvent.click(testConfigBtn)
    })
    await waitFor(() => expect(getByText('authSettings.ldap.verifyAndEnableConfig')).not.toBeNull())
    const enableLdapConfigBtn = getByTestId('enable-ldap-config')
    await act(async () => {
      enableLdapConfigBtn && fireEvent.click(enableLdapConfigBtn)
    })
    const emailField = document.querySelector('[name="email"]')
    const passwordField = document.querySelector('[name="password"]')
    await act(async () => {
      emailField && fireEvent.change(emailField, { target: { value: 'test@harness.io' } })
      passwordField && fireEvent.change(passwordField, { target: { value: 'test@1234' } })
    })
    await act(async () => {
      enableLdapConfigBtn && fireEvent.click(enableLdapConfigBtn)
    })
    await waitFor(() => expect(getByText('Not Allowed')).not.toBeNull())
  })
  test('LDAP Provider test config', async () => {
    jest.spyOn(cdngServices, 'usePostLdapLoginTest').mockImplementation(
      () =>
        ({
          mutate: mockLdapLoginTest
        } as any)
    )
    const { getByTestId, getByText } = render(
      <TestWrapper pathParams={{ accountId: 'testAcc' }} defaultFeatureFlagValues={{ NG_ENABLE_LDAP_CHECK: true }}>
        <LDAPProvider
          authSettings={mockAuthSettingsResponse as AuthenticationSettingsResponse}
          canEdit={true}
          refetchAuthSettings={refetchAuthSettings}
          permissionRequest={permissionRequest}
          setUpdating={mockDispatch}
        />
      </TestWrapper>
    )

    const testConfigBtn = getByTestId('ldap-config-test')
    act(() => {
      testConfigBtn && fireEvent.click(testConfigBtn)
    })
    await waitFor(() => expect(getByText('authSettings.ldap.verifyAndEnableConfig')).not.toBeNull())
    const testLdapConfigBtn = getByTestId('test-ldap-config')

    const emailField = document.querySelector('[name="email"]')
    const passwordField = document.querySelector('[name="password"]')
    await act(async () => {
      emailField && fireEvent.change(emailField, { target: { value: 'test@harness.io' } })
      passwordField && fireEvent.change(passwordField, { target: { value: 'test@1234' } })
    })
    await act(async () => {
      testLdapConfigBtn && fireEvent.click(testLdapConfigBtn)
    })
    await waitFor(() => expect(getByText('authSettings.ldap.ldapTestSuccessful')).not.toBeNull())
  })
  test('LDAP Provider test config fail', async () => {
    jest.spyOn(cdngServices, 'usePostLdapLoginTest').mockImplementation(
      () =>
        ({
          mutate: mockLdapLoginTestFail
        } as any)
    )
    const { getByTestId, getByText } = render(
      <TestWrapper pathParams={{ accountId: 'testAcc' }} defaultFeatureFlagValues={{ NG_ENABLE_LDAP_CHECK: true }}>
        <LDAPProvider
          authSettings={mockAuthSettingsResponse as AuthenticationSettingsResponse}
          canEdit={true}
          refetchAuthSettings={refetchAuthSettings}
          permissionRequest={permissionRequest}
          setUpdating={mockDispatch}
        />
      </TestWrapper>
    )

    const testConfigBtn = getByTestId('ldap-config-test')
    act(() => {
      testConfigBtn && fireEvent.click(testConfigBtn)
    })
    await waitFor(() => expect(getByText('authSettings.ldap.verifyAndEnableConfig')).not.toBeNull())
    const testLdapConfigBtn = getByTestId('test-ldap-config')

    const emailField = document.querySelector('[name="email"]')
    const passwordField = document.querySelector('[name="password"]')
    await act(async () => {
      emailField && fireEvent.change(emailField, { target: { value: 'test@harness.io' } })
      passwordField && fireEvent.change(passwordField, { target: { value: 'test@1234' } })
    })
    await act(async () => {
      testLdapConfigBtn && fireEvent.click(testLdapConfigBtn)
    })
    await waitFor(() => expect(getByText('Invalid Credentials')).not.toBeNull())
  })
  test('LDAP Provider group sync', async () => {
    jest.spyOn(cdngServices, 'syncLdapGroupsPromise').mockImplementation(() => mockGroupSync())
    const { getByTestId, getByText } = render(
      <TestWrapper pathParams={{ accountId: 'testAcc' }} defaultFeatureFlagValues={{ NG_ENABLE_LDAP_CHECK: true }}>
        <LDAPProvider
          authSettings={mockAuthSettingsResponse as AuthenticationSettingsResponse}
          canEdit={true}
          refetchAuthSettings={refetchAuthSettings}
          permissionRequest={permissionRequest}
          setUpdating={mockDispatch}
        />
      </TestWrapper>
    )
    const ldapPopoverEl = getByTestId('provider-button')
    expect(ldapPopoverEl).not.toBeNull()
    act(() => {
      ldapPopoverEl && fireEvent.click(ldapPopoverEl)
    })

    const triggerGroupSyncBtn = getByTestId('ldap-group-sync')
    expect(triggerGroupSyncBtn).not.toBeNull()
    act(() => {
      triggerGroupSyncBtn && fireEvent.click(triggerGroupSyncBtn)
    })
    await waitFor(() => expect(getByText('authSettings.ldap.syncUserGroupsResult.fail')).not.toBeNull())
  })
})
