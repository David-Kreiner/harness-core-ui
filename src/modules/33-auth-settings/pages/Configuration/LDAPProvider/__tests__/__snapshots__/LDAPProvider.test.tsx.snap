// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LDAP Provider LDAP Provider Snapshot 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main StyledProps--margin-xlarge StyledProps--background-white"
  >
    <div
      class="Collapse--main collapseClassName"
    >
      <div
        class="StyledProps--font StyledProps--main CollapseHeader--main Collapse--header collapseHeaderClassName height60"
      >
        <div
          class="StyledProps--font StyledProps--main CollapseHeader--leftSection"
        >
          <span
            class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-grey400"
            data-icon="main-chevron-up"
          >
            <svg
              height="10"
              viewBox="0 0 14 8"
              width="10"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M13.7 6.3l-6-6C7.5.1 7.3 0 7 0s-.5.1-.7.3l-6 6c-.2.2-.3.4-.3.7 0 .6.4 1 1 1 .3 0 .5-.1.7-.3L7 2.4l5.3 5.3c.2.2.4.3.7.3.6 0 1-.4 1-1 0-.3-.1-.5-.3-.7z"
                fill="#4d5969"
              />
            </svg>
          </span>
          <div
            class="StyledProps--font StyledProps--main StyledProps--margin-left-xlarge"
          >
            <label
              class="bp3-control bp3-radio StyledProps--font Radio--radio StyledProps--main StyledProps--font-weight-bold StyledProps--font-size-normal StyledProps--color StyledProps--color-grey900"
            >
              <input
                checked=""
                type="radio"
              />
              <span
                class="bp3-control-indicator"
              />
              authSettings.loginViaLDAP
            </label>
          </div>
        </div>
        <div
          class="StyledProps--font StyledProps--main"
        />
      </div>
      <div
        class="Collapse--collapse"
      >
        <div
          class="bp3-collapse"
          style="height: auto; overflow-y: visible; transition: none;"
        >
          <div
            aria-hidden="false"
            class="bp3-collapse-body"
            style="transform: translateY(0); transition: none;"
          >
            <div
              class="StyledProps--font StyledProps--main StyledProps--padding-bottom-large"
            >
              <div
                class="bp3-card bp3-elevation-0 Card--card card"
              >
                <div
                  class="StyledProps--font StyledProps--main StyledProps--margin-left-xlarge"
                >
                  <label
                    class="bp3-control bp3-radio StyledProps--font Radio--radio StyledProps--main StyledProps--font-weight-bold StyledProps--font-size-normal"
                  >
                    <input
                      checked=""
                      readonly=""
                      type="radio"
                    />
                    <span
                      class="bp3-control-indicator"
                    />
                  </label>
                </div>
                <p
                  class="StyledProps--font StyledProps--main flexGrowFull StyledProps--font-variation-card-title"
                >
                  ldap11223355
                </p>
                <button
                  class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--variation Button--variation-secondary"
                  data-testid="ldap-config-test"
                  type="button"
                >
                  <span
                    class="bp3-button-text"
                  >
                    test
                  </span>
                </button>
                <span
                  class="bp3-popover-wrapper"
                >
                  <span
                    class="bp3-popover-target"
                  >
                    <button
                      class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly Button--without-shadow Button--withLeftIcon Button--variation Button--variation-icon"
                      data-testid="provider-button"
                      type="button"
                    >
                      <span
                        class="bp3-icon StyledProps--main"
                        data-icon="Options"
                      >
                        <svg
                          fill="currentColor"
                          height="16"
                          viewBox="0 0 4 16"
                          width="16"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <circle
                            cx="2"
                            cy="2"
                            r="1.5"
                          />
                          <circle
                            cx="2"
                            cy="8"
                            r="1.5"
                          />
                          <circle
                            cx="2"
                            cy="14"
                            r="1.5"
                          />
                        </svg>
                      </span>
                    </button>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`LDAP Provider LDAP Provider Snapshot with no data 1`] = `
<div>
  <div
    class="StyledProps--font StyledProps--main StyledProps--margin-xlarge StyledProps--background-white"
  >
    <div
      class="bp3-card bp3-elevation-0 Card--card cardWithRadioBtn"
    >
      <div
        class="StyledProps--font StyledProps--main StyledProps--margin-left-xlarge StyledProps--margin-top-xsmall"
      >
        <label
          class="bp3-control bp3-radio StyledProps--font Radio--radio StyledProps--main StyledProps--font-weight-semi-bold StyledProps--font-size-normal StyledProps--color StyledProps--color-primary7"
        >
          <input
            type="radio"
          />
          <span
            class="bp3-control-indicator"
          />
          authSettings.ldap.addLdap
        </label>
      </div>
    </div>
  </div>
</div>
`;
